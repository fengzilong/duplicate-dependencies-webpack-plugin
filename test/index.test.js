const path = require( 'path' )
const webpack = require( 'webpack' )
const Plugin = require( '../' )
const DuplicateReadPlugin = require( './duplicate-read-plugin' )

describe( 'basic', () => {
  test( 'snapshot', async done => {
    const compiler = getCompiler( path.join( __dirname, 'fixtures/entry.js' ) )
    
    new Plugin().apply( compiler )
    new DuplicateReadPlugin( {
      callback( duplicates ) {
        formatDuplicates( duplicates )
        expect( duplicates ).toMatchSnapshot()
        done()
      }
    } ).apply( compiler )
    
    await run( compiler )
  } )
} )

function getCompiler( entry ) {
  const compiler = webpack( {
    entry,
    mode: 'development',
  } )
  
  const mfs = new webpack.MemoryOutputFileSystem()

  compiler.outputFileSystem = mfs
  
  return compiler
}

function run( compiler ) {
  return new Promise( ( resolve, reject ) => {
    compiler.run( ( error, stats ) => {
      if ( error ) {
        return reject( error )
      }

      return resolve( stats )
    } )
  } )
}

function formatDuplicates( duplicates = {} ) {
  const cwd = process.cwd()

  Object.keys( duplicates )
    .forEach( name => {
      const versions = duplicates[ name ] || []
      versions.forEach( version => {
        version.issuer = path.relative( cwd, version.issuer )
        version.root = path.relative( cwd, version.root )
        version.files = version.files.map( file => {
          return path.relative( cwd, file )
        } )
      } )
    } )
}
