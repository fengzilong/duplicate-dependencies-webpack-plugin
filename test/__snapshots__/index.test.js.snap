// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic snapshot 1`] = `
Object {
  "deepmerge": Array [
    Object {
      "bytes": 4035,
      "files": Array [
        "node_modules/deepmerge/dist/cjs.js",
      ],
      "issuer": "test/fixtures/entry.js",
      "issuerPath": Array [
        "test/fixtures/entry.js",
      ],
      "name": "deepmerge",
      "root": "node_modules/deepmerge",
      "version": "4.2.2",
    },
    Object {
      "bytes": 3309,
      "files": Array [
        "node_modules/element-ui/node_modules/deepmerge/dist/cjs.js",
      ],
      "issuer": "node_modules/element-ui/lib/locale/index.js",
      "issuerPath": Array [
        "node_modules/element-ui/lib/locale/index.js",
        "node_modules/element-ui/lib/element-ui.common.js",
        "test/fixtures/entry.js",
      ],
      "name": "deepmerge",
      "root": "node_modules/element-ui/node_modules/deepmerge",
      "version": "1.5.2",
    },
  ],
}
`;
